pantahub/pantahub_proto.c

```diff
@@ -0,0 +1,211 @@
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+#include <string.h>
+
+#include <jsmn/jsmnutil.h>
+
+#include "pantahub/pantahub_proto.h"
+#include "pantahub/pantahub_rest.h"
+
+#include "metadata.h"
+
+#include "utils/json.h"
+#include "utils/str.h"
+
+#define MODULE_NAME "pantahub_proto"
+#define pv_log(level, msg, ...) vlog(MODULE_NAME, level, msg, ##__VA_ARGS__)
+#include "log.h"
+
+typedef struct {
+	char *token;
+} pantahub_session_t;
+
+static pantahub_session_t session = { 0 };
+
+static void _parse_post_auth_body(const char *json)
+{
+	int tokc;
+	jsmntok_t *tokv = NULL;
+
+	if (jsmnutil_parse_json(json, &tokv, &tokc) < 0) {
+		goto out;
+	}
+
+	session.token = pv_json_get_value(json, "token", tokv, tokc);
+
+out:
+	if (tokv)
+		free(tokv);
+}
+
+static void _recv_post_auth_cb(struct evhttp_request *req, void *ctx)
+{
+	pv_log(DEBUG, "POST auth response received");
+
+	char buffer[1024];
+	memset(&buffer, 0, sizeof(buffer));
+	if (pv_pantahub_rest_recv(req, ctx, &buffer[0], 1024) == 200)
+		_parse_post_auth_body(buffer);
+
+	if (session.token)
+		pv_log(DEBUG, "new session opened");
+}
+
+void pv_pantahub_proto_open_session(struct event_base *base)
+{
+	if (session.token) {
+		pv_log(ERROR, "session must be closed first");
+		return;
+	}
+
+	pv_log(DEBUG, "opening new session...");
+
+	char uri[256];
+	SNPRINTF_WTRUNC(uri, sizeof(uri), "/auth/login");
+
+	struct pv_json_ser js;
+	pv_json_ser_init(&js, 1024);
+
+	pv_json_ser_object(&js);
+	{
+		pv_json_ser_key(&js, "username");
+		pv_json_ser_string(&js, pv_config_get_str(PH_CREDS_PRN));
+		pv_json_ser_key(&js, "password");
+		pv_json_ser_string(&js, pv_config_get_str(PH_CREDS_SECRET));
+		pv_json_ser_object_pop(&js);
+	}
+
+	char *body = pv_json_ser_str(&js);
+	if (!body) {
+		pv_log(ERROR, "could not serialize session auth JSON");
+		return;
+	}
+
+	pv_pantahub_rest_send(base, EVHTTP_REQ_POST, uri, NULL, body,
+			      _recv_post_auth_cb);
+	free(body);
+}
+
+bool pv_pantahub_proto_is_session_open()
+{
+	return session.token;
+}
+
+void pv_pantahub_proto_close_session()
+{
+	if (session.token)
+		free(session.token);
+	session.token = NULL;
+
+	pv_log(DEBUG, "session closed");
+}
+
+typedef struct {
+	char *usrmeta;
+	char *devmeta;
+} pantahub_cache_t;
+
+static pantahub_cache_t cache = { 0 };
+
+static bool _is_in_cache(const char *in, char *cache)
+{
+	return cache && in && pv_str_matches(cache, strlen(cache), in, strlen(in));
+}
+
+static void _save_cache(const char *in, char **cache)
+{
+	if (*cache)
+		free(*cache);
+	*cache = strdup(in);
+}
```

Comment:
this cache concept is not clear to me.... what is it and should we really have this globally rater than inside pv?

Reply:
I ended up hanging it from a new pantahub_t global similar to pv struct. I removed user meta cache, which is redundant, and left devmeta cache so we know when to PUT. Let me know if you don't like it.